### PR TITLE
Generate JSON types for entql

### PR DESCRIPTION
**Pull request by ivanvanderbyl**

### CODE REVIEW

**Comment by a8m** [comment]
Hey @ivanvanderbyl and thanks for your contribution.
Do we really need to generate a new predicate-type for JSON fields? can't we use the `[]byte` predicates instead?

**Comment by ivanvanderbyl** [comment]
@a8m good question — I suppose that would work. I was more concerned with fixing the code gen for entql in places I already have JSON types. But the underlying predicates aren't important to me.

**Comment by ivanvanderbyl** [comment]
@a8m would it make sense to implement only Eq and Neq predicates for JSON?

**Comment by BrentChesny** [comment]
Just hit this same issue. The exact predicates that are available don't matter too much for me as long as the codegen produces buildable code. :)

**Comment by a8m** [comment]
>  The exact predicates that are available don't matter too much for me as long as the codegen produces buildable code. :)

@ivanvanderbyl, do you want to try adding the JSON predicates using `entql.BytesP`?

**Comment by ivanvanderbyl** [comment]
I'll give that a try later today

**Comment by ivanvanderbyl** [comment]
How's this look @a8m ?

File: entc/gen/template/dialect/sql/entql.tmpl [hunk 1]
```diff
@@ -118,7 +118,11 @@ type predicateAdder interface {
 	{{ range $f := $n.Fields }}
 		{{ $type := $f.Type.Type.String }}
 		{{ $iface := print (pascal $type) "P" }}
-		{{ if $f.IsTime }}{{ $iface = "TimeP" }}{{ else if $f.IsBytes }}{{ $iface = "BytesP" }}{{ else if $f.IsUUID }}{{ $iface = "ValueP" }}{{ end }}
+		{{- if $f.IsTime }}{{ $iface = "TimeP" }}
+		{{- else if $f.IsBytes }}{{ $iface = "BytesP" }}
+		{{- else if $f.IsUUID }}{{ $iface = "ValueP" }}
+		{{- else if $f.IsJSON }}{{ $iface = "BytesP" }}
```
Comment (a8m) [inline entc/gen/template/dialect/sql/entql.tmpl:124]:
```suggestion
		{{- else if or $f.IsBytes $f.IsJSON }}{{ $iface = "BytesP" }}
		{{- else if $f.IsUUID }}{{ $iface = "ValueP" }}
```

File: entql/internal/gen.go [hunk 2]
```diff
@@ -74,7 +74,7 @@ func ops(t field.Type) []string {
 
 func ident(t field.Type) string {
 	switch t {
-	case field.TypeBytes:
+	case field.TypeBytes, field.TypeJSON:
```
Comment (a8m) [inline entql/internal/gen.go:77]:
```suggestion
	case field.TypeBytes:
```
I'm pretty sure this one is not needed.

File: entc/gen/type.go [hunk 3]
```diff
@@ -895,8 +895,7 @@ func (f Field) NullTypeField(rec string) string {
 	return expr
 }
 
-// Column returns the table column. It sets it as a primary key (auto_increment) in case of ID field, unless stated
-// otherwise.
+// Column returns the table column. It sets it as a primary key (auto_increment) in case of ID field.
```
Comment (a8m) [inline entc/gen/type.go:898]: Can you return back the previous comment (rebase/merge issue)?

**Review by a8m — APPROVED** [review]
Looks good @ivanvanderbyl, please fix these small comments and I'll merge this. Thanks for your contribution!

**Comment by a8m** [comment]
Hey @ivanvanderbyl, CI is failing. Please run `go generate ./entc/gen/...` and then `go generate ./...` from the root directory.

**Review by a8m — APPROVED** [review]

**Comment by a8m** [comment]
I created this PR for merging this - https://github.com/facebook/ent/pull/1112

**Comment by a8m** [comment]
Merged #1112 - Thanks a lot @ivanvanderbyl

**Pull request closed without merging**